datumaro/plugins/ndr.py

```diff
@@ -0,0 +1,260 @@
+            num_cut=None, over_sample='random', under_sample='uniform', seed=None, **kwargs):
+        """
+        Near-duplicated image removal
+
+        Arguments
+        ---------------
+        working_subset: str
+            name of the subset to operate
+            if None, use DEFAULT_SUBSET_NAME
+        duplicated_subset: str
+            name of the subset for the removed data after NDR runs
+        algorithm: str
+            name of the algorithm to use
+            "gradient" only for now.
+        num_cut: int
+            number of outputs you want.
+            the algorithm will cut whole dataset to this amount
+            if None, return result without any modification
+        over_sample: "random" or "similarity"
+            specify the strategy when num_cut > length of the result after removal
+            if random, sample from removed data randomly
+            if similarity, select from removed data with ascending order of similarity
+        under_sample: "uniform" or "inverse"
+            specify the strategy when num_cut < length of the result after removal
+            if uniform, sample data with uniform distribution
+            if inverse, sample data with reciprocal of the number of data which have same hash key
+        Algorithm Specific for gradient
+            block_shape: tuple, (h, w)
+                for the robustness, this function will operate on blocks
+                mean and variance will be calculated on this block
+            hash_dim: int
+                dimension(or bit) of the hash function
+            sim_threshold: float
+                the threshold value for saving hash-collided samples.
+                larger value means more generous, i.e., saving more samples
+        Return
+        ---------------
+        None, other subsets combined with the result
+        """
+        super().__init__(extractor)
+        if seed:
+            np.random.seed(seed)
+        if len(extractor) < 1:
+            raise ValueError("Empty dataset")
+        if not working_subset:
+            working_subset = DEFAULT_SUBSET_NAME
+        if working_subset not in extractor.subsets():
+            raise ValueError("Invalid working_subset name")
+        self.working_subset = working_subset
+
+        # parameter validation before main runs
+        if working_subset == duplicated_subset:
+            raise ValueError("working_subset == duplicated_subset")
+        if algorithm not in ('gradient'):
+            raise ValueError("Invalid algorithm name")
+        if over_sample not in ("random", "similarity"):
+            raise ValueError("Invalid over_sample")
+        if under_sample not in ("uniform", "inverse"):
+            raise ValueError("Invalid under_sample")
+        
+        self.duplicated_subset = duplicated_subset
+        self._remove(algorithm, working_subset, num_cut, over_sample, under_sample, **kwargs)
```

Comment:
Try to make the transform lazy - it should only be applied when `__iter__` is called, look at "splitter" transforms for example. You'll also need to move the initialization of RNG into `__iter__`. Consider `__iter__` in the input dataset a heavy operation.

Reply:
Okay, I will check "splitter" and change to lazy transformation